substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/traceid/JfrTraceId.java

```diff
@@ -0,0 +1,157 @@
+
+    private static final int TRACE_ID_SHIFT = 16;
+
+    // Epoch stuff
+    private static final long USED_BIT = 1;
+    private static final int EPOCH_1_SHIFT = 0;
+    private static final int EPOCH_2_SHIFT = 1;
+    private static final long USED_EPOCH_1_BIT = USED_BIT << EPOCH_1_SHIFT;
+    private static final long USED_EPOCH_2_BIT = USED_BIT << EPOCH_2_SHIFT;
+
+    private static final long JDK_JFR_EVENT_SUBKLASS = 16;
+    private static final long JDK_JFR_EVENT_KLASS = 32;
+    private static final long EVENT_HOST_KLASS = 64;
+
+    @Fold
+    static JfrTraceIdMap getTraceIdMap() {
+        return ImageSingletons.lookup(JfrTraceIdMap.class);
+    }
+
+    public static void tag(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        map.setId(clazz, (id & ~0xff) | (bits & 0xff));
+    }
+
+    public static boolean predicate(Class<?> clazz, long bits) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(clazz);
+        return (id & bits) != 0;
+    }
+
+    @Uninterruptible(reason = "Epoch must not change.")
+    public static void setUsedThisEpoch(Class<?> clazz) {
+        tag(clazz, JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    @Uninterruptible(reason = "Epoch must not change.")
+    public static boolean isUsedThisEpoch(Class<?> clazz) {
+        return predicate(clazz, TRANSIENT_BIT | JfrTraceIdEpoch.getInstance().thisEpochBit());
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static long getTraceIdRaw(Class<?> clazz) {
+        return getTraceIdMap().getId(clazz);
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    public static long getTraceId(Class<?> clazz) {
+        long traceid = getTraceIdRaw(clazz);
+        return traceid >>> TRACE_ID_SHIFT;
+    }
+
+    public static long load(Class<?> clazz) {
+        return JfrTraceIdLoadBarrier.load(clazz);
+    }
+
+    private static void tagAsJdkJfrEvent(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        map.setId(index, id | JDK_JFR_EVENT_KLASS);
+    }
+
+    private static void tagAsJdkJfrEventSub(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        map.setId(index, id | JDK_JFR_EVENT_SUBKLASS);
+    }
+
+    private static boolean isEventClass(int index) {
+        JfrTraceIdMap map = getTraceIdMap();
+        long id = map.getId(index);
+        return (id & (JDK_JFR_EVENT_KLASS | JDK_JFR_EVENT_SUBKLASS)) != 0;
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public static void assign(Class<?> clazz, int index) {
+        assert clazz != null;
+        if (getTraceIdMap().getId(index) != -1) {
+            return;
+        }
+        long typeId = JVM.getJVM().getTypeId(clazz);
```

Comment:
Here you asking HotSpot during the native image build for a JFR type id. Is this necessary to keep things in sync or could we just use the id from the `DynamicHub` instead?

Reply:
Yes, this is one main point why we need to do all this dance. Hotspot's JFR code - the Java parts - are building some datastructures themselves which use Hotspot's trace-IDs, and which get built into the image. We need to be in sync with them.

Reply:
Please add a comment that mentions that.